File: SuperiorPackGroup/Code Classes/LotCodeValidator.vb
Imports DevExpress.Xpo
Imports DXDAL.SPGData

Public Class LotCodeValidator

    Public Shared Function ValidateByItem(ByVal item As Items, ByVal lot As String, ByVal ValidateFormat As Boolean) As Boolean
        Dim sortCollection As SortingCollection = New SortingCollection()
        Dim codeSection As LotCodeFormatSections
        Dim sectionSettings As LotCodeSectionSettings
        Dim sectionValue As String, sectionValueNum As Integer, valuePosition As Integer = 0

        If Not If(item.RequiresLotCodes, False) Then Return True

        If IsNothing(lot) OrElse lot = "" Then Return False

        If Not ValidateFormat Then Return True

        If item.DefaultLotCodeFormat Is Nothing Then Return True

        sortCollection.Add(New SortProperty("SectionPosition", DB.SortingDirection.Ascending))
        item.DefaultLotCodeFormat.FormatSections.Sorting = sortCollection

        For Each codeSection In item.DefaultLotCodeFormat.FormatSections

            If lot.Length < valuePosition + codeSection.SectionLength Then
                Return False
            End If

            sectionValue = Mid(lot, valuePosition + 1, codeSection.SectionLength)

            For Each sectionSettings In codeSection.SectionSettings
                If sectionSettings.SettingName = "NumberRangeFrom" Then
                    If Not Integer.TryParse(sectionValue, sectionValueNum) OrElse sectionValueNum < CInt(sectionSettings.SettingValue) Then Return False
                End If
                If sectionSettings.SettingName = "NumberRangeTo" Then
                    If Not Integer.TryParse(sectionValue, sectionValueNum) OrElse sectionValueNum > CInt(sectionSettings.SettingValue) Then Return False
                End If
            Next

            valuePosition += codeSection.SectionLength

        Next

        Return True
    End Function
    Public Shared Function ValidateByItemID(ByVal itemId As Integer, ByVal lot As String, ByVal ValidateFormat As Boolean) As Boolean
        Dim item As Items
        item = Session.DefaultSession.GetObjectByKey(Of Items)(itemId)

        If item Is Nothing Then Return False

        Return ValidateByItem(item, lot, ValidateFormat)

    End Function



End Class

File: SuperiorPackGroup/Code Classes/LotCodeValidator.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;

using DevExpress.Xpo;
using DXDAL.SPGData;

namespace SuperiorPackGroup
{
	public class LotCodeValidator
	{

		public static bool ValidateByItem(Items item, string lot, bool ValidateFormat)
		{
			SortingCollection sortCollection = new SortingCollection();
			LotCodeFormatSections codeSection = null;
//INSTANT C# NOTE: Commented this declaration since looping variables in 'foreach' loops are declared in the 'foreach' header in C#:
//			Dim sectionSettings As LotCodeSectionSettings
			string sectionValue = null;
			int sectionValueNum = 0;
			int valuePosition = 0;

			if (!(item.RequiresLotCodes ?? false))
			{
				return true;
			}

			if ((lot == null) || string.IsNullOrEmpty(lot))
			{
				return false;
			}

			if (!ValidateFormat)
			{
				return true;
			}

			if (item.DefaultLotCodeFormat == null)
			{
				return true;
			}

			sortCollection.Add(new SortProperty("SectionPosition", DevExpress.Xpo.DB.SortingDirection.Ascending));
			item.DefaultLotCodeFormat.FormatSections.Sorting = sortCollection;

			foreach (LotCodeFormatSections codeSectionWithinLoop in item.DefaultLotCodeFormat.FormatSections)
			{
			codeSection = codeSectionWithinLoop;

				if (lot.Length < valuePosition + codeSectionWithinLoop.SectionLength)
				{
					return false;
				}

				sectionValue = lot.Substring(valuePosition, codeSectionWithinLoop.SectionLength);

				foreach (LotCodeSectionSettings sectionSettings in codeSectionWithinLoop.SectionSettings)
				{
					if (sectionSettings.SettingName == "NumberRangeFrom")
					{
						if (!int.TryParse(sectionValue, out sectionValueNum) || sectionValueNum < int.Parse(sectionSettings.SettingValue))
						{
							return false;
						}
					}
					if (sectionSettings.SettingName == "NumberRangeTo")
					{
						if (!int.TryParse(sectionValue, out sectionValueNum) || sectionValueNum > int.Parse(sectionSettings.SettingValue))
						{
							return false;
						}
					}
				}

				valuePosition += codeSectionWithinLoop.SectionLength;

			}

			return true;
		}
		public static bool ValidateByItemID(int itemId, string lot, bool ValidateFormat)
		{
			Items item = Session.DefaultSession.GetObjectByKey<Items>(itemId);

			if (item == null)
			{
				return false;
			}

			return ValidateByItem(item, lot, ValidateFormat);

		}



	}

}